File: imagepicker/src/main/java/com/rxxb/imagepicker/ui/CropActivity.java


package com.rxxb.imagepicker.ui;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.support.annotation.Nullable;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.TextView;
import com.jw.uilibrary.base.activity.BaseActivity;
import com.rxxb.imagepicker.ImagePicker;
import com.rxxb.imagepicker.R;
import com.rxxb.imagepicker.R.id;
import com.rxxb.imagepicker.R.string;
import com.rxxb.imagepicker.crop.CropIwaView;
import com.rxxb.imagepicker.crop.CropIwaView.CropSaveCompleteListener;
import com.rxxb.imagepicker.crop.CropIwaView.ErrorListener;
import com.rxxb.imagepicker.crop.config.CropIwaSaveConfig.Builder;
import com.rxxb.imagepicker.crop.shape.CropIwaOvalShape;
import com.rxxb.imagepicker.view.CropImageView.Style;

import java.io.File;

import static com.rxxb.imagepicker.ImagePicker.EXTRA_CROP_IMAGE_OUT_URI;

public class CropActivity extends ImageBaseActivity implements OnClickListener {
    private static final String EXTRA_URI = "CropImage";
    private ImagePicker imagePicker;
    private CropIwaView cropView;
    private ProgressDialog mProgressDialog;
    private String dstPath;
    private float originAngle;

    public CropActivity() {
    }

    public static Intent callingIntent(Context context, Uri imageUri) {
        Intent intent = new Intent(context, CropActivity.class);
        intent.putExtra("CropImage", imageUri);
        return intent;
    }

    @Override
    public void doInflate(BaseActivity activity, Bundle savedInstanceState) {
        super.doInflate(activity, savedInstanceState);
    }

    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.setContentView(R.layout.activity_crop);
        this.findViewById(id.tv_rotate).setOnClickListener(this);
        this.findViewById(id.tv_recover).setOnClickListener(this);
        this.findViewById(id.btn_back).setOnClickListener(this);
        Button mBtnOk = (Button) this.findViewById(id.btn_ok);
        mBtnOk.setText(this.getString(string.ip_complete));
        mBtnOk.setOnClickListener(this);
        this.imagePicker = ImagePicker.getInstance();
        Uri imageUri = (Uri) this.getIntent().getParcelableExtra("CropImage");
        this.cropView = (CropIwaView) this.findViewById(id.cv_crop_image);
        this.cropView.setImageUri(imageUri);
        this.cropView.configureOverlay().setAspectRatio(this.imagePicker.getAspectRatio()).setDynamicCrop(this.imagePicker.isDynamicCrop()).apply();
        if (this.imagePicker.getStyle() == Style.CIRCLE) {
            this.cropView.configureOverlay().setCropShape(new CropIwaOvalShape(this.cropView.configureOverlay())).apply();
        }

        File cropCacheFolder;
        if (this.imagePicker.getCutType() == 2) {
            cropCacheFolder = new File(Environment.getExternalStorageDirectory() + "/RXImagePicker/");
        } else {
            cropCacheFolder = this.imagePicker.getCropCacheFolder(this);
        }

        if (!cropCacheFolder.exists() || !cropCacheFolder.isDirectory()) {
            cropCacheFolder.mkdirs();
        }

        this.dstPath = (new File(cropCacheFolder, "IMG_" + System.currentTimeMillis() + ".png")).getAbsolutePath();
        this.cropView.setCropSaveCompleteListener(new CropSaveCompleteListener() {
            public void onCroppedRegionSaved(Uri bitmapUri) {
                CropActivity.this.dismiss();
                ImagePicker.galleryAddPic(CropActivity.this.getApplicationContext(), bitmapUri);
                Intent intent = new Intent();
                intent.putExtra(EXTRA_CROP_IMAGE_OUT_URI, bitmapUri);
                CropActivity.this.setResult(-1, intent);
                CropActivity.this.finish();
            }
        });
        this.cropView.setErrorListener(new ErrorListener() {
            public void onError(Throwable e) {
                CropActivity.this.dismiss();
            }
        });
        this.originAngle = this.cropView.getMatrixAngle();
        this.setConfirmButtonBg(mBtnOk);
        this.findViewById(id.top_bar).setBackgroundColor(Color.parseColor(this.imagePicker.getViewColor().getNaviBgColor()));
        ((TextView) this.findViewById(id.tv_des)).setTextColor(Color.parseColor(this.imagePicker.getViewColor().getNaviTitleColor()));
    }

    private void dismiss() {
        if (this.mProgressDialog != null && this.mProgressDialog.isShowing()) {
            this.mProgressDialog.dismiss();
        }

    }

    public void onClick(View v) {
        int id = v.getId();
        if (id == R.id.btn_back) {
            this.setResult(0);
            this.finish();
        } else if (id == R.id.tv_rotate) {
            this.cropView.rotateImage(90.0F);
        } else if (id == R.id.tv_recover) {
            float currentAngle = this.cropView.getMatrixAngle();
            if (this.originAngle != currentAngle) {
                this.cropView.rotateImage(this.originAngle - currentAngle);
            }

            this.cropView.initialize();
        } else if (id == R.id.btn_ok) {
            if (this.mProgressDialog == null) {
                this.mProgressDialog = new ProgressDialog(this);
                this.mProgressDialog.setMessage("正在处理中...");
                this.mProgressDialog.setCanceledOnTouchOutside(false);
                this.mProgressDialog.setCancelable(false);
            }

            this.mProgressDialog.show();
            Builder builder = new Builder(Uri.fromFile(new File(this.dstPath)));
            if (this.imagePicker.getOutPutX() != 0 && this.imagePicker.getOutPutY() != 0 && !this.imagePicker.isOrigin()) {
                builder.setSize(this.imagePicker.getOutPutX(), this.imagePicker.getOutPutY());
            }

            builder.setQuality(this.imagePicker.getQuality());
            this.cropView.crop(builder.build());
        }

    }

    protected void onDestroy() {
        super.onDestroy();
        this.dismiss();
        this.mProgressDialog = null;
    }
}

File: croplibrary/src/main/java/com/jw/croplibrary/img/CropActivity.kt
package com.jw.croplibrary.img

import android.app.Activity
import android.app.ProgressDialog
import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View.OnClickListener
import com.jw.croplibrary.CropLibrary
import com.jw.croplibrary.R
import com.jw.croplibrary.databinding.ActivityCropBinding
import com.jw.croplibrary.img.config.CropIwaSaveConfig.Builder
import com.jw.croplibrary.img.shape.CropIwaOvalShape
import com.jw.library.ColorCofig
import com.jw.library.ui.BaseBindingActivity
import com.jw.library.utils.BitmapUtil
import com.jw.library.utils.FileUtils
import java.io.File

/**
 * 创建时间：
 * 更新时间
 * 版本：
 * 作者：Mr.jin
 * 描述：图片裁剪Activity
 */
class CropActivity : BaseBindingActivity<ActivityCropBinding>(),
    CropIwaView.CropSaveCompleteListener, CropIwaView.ErrorListener {
    private var mProgressDialog: ProgressDialog? = null
    private var dstPath: String? = null
    private var originAngle: Float = 0.toFloat()

    override fun getLayoutId() = R.layout.activity_crop

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        releaseFolder()
        mBinding.apply {
            clickListener = OnClickListener {
                when (it.id) {
                    R.id.tv_rotate -> rotate()
                    R.id.tv_recover -> recover()
                    R.id.btn_back -> finish()
                    R.id.btn_ok -> crop()
                }
            }
            topBar.btnOk.text = getString(R.string.ip_complete)
            setConfirmButtonBg(topBar.btnOk)
        }

        val imageUri = intent.getParcelableExtra<Uri>("CropImage")
        mBinding.cvCropImage.setImageUri(imageUri)
        mBinding.cvCropImage.configureOverlay().setAspectRatio(CropLibrary.aspectRatio)
            .setDynamicCrop(CropLibrary.isDynamicCrop).apply()
        if (CropLibrary.style == CropImageView.Style.CIRCLE) {
            mBinding.cvCropImage.configureOverlay()
                .setCropShape(CropIwaOvalShape(mBinding.cvCropImage.configureOverlay())).apply()
        }
        dstPath =
            File(
                CropLibrary.CACHE_IMG_CROP,
                "IMG_" + System.currentTimeMillis() + ".png"
            ).absolutePath
        mBinding.cvCropImage.setCropSaveCompleteListener(this)
        mBinding.cvCropImage.setErrorListener { this@CropActivity.dismiss() }
        originAngle = mBinding.cvCropImage.matrixAngle
        mBinding.topBar.tvDes.setTextColor(Color.parseColor(ColorCofig.naviTitleColor))
    }

    private fun rotate() {
        mBinding.cvCropImage.rotateImage(90.0f)
    }

    private fun recover() {
        val currentAngle = mBinding.cvCropImage.matrixAngle
        if (originAngle != currentAngle) {
            mBinding.cvCropImage.rotateImage(originAngle - currentAngle)
        }

        mBinding.cvCropImage.initialize()
    }

    private fun crop() {
        if (mProgressDialog == null) {
            mProgressDialog = ProgressDialog(this@CropActivity)
            mProgressDialog!!.setMessage("正在处理中...")
            mProgressDialog!!.setCanceledOnTouchOutside(false)
            mProgressDialog!!.setCancelable(false)
        }

        mProgressDialog!!.show()
        val builder = Builder(Uri.fromFile(File(dstPath)))
/*        if (CropLibrary.outPutX != 0 && CropLibrary.outPutY != 0) {
            builder.setSize(CropLibrary.outPutX, CropLibrary.outPutY)
        }*/

        builder.setQuality(CropLibrary.quality)
        mBinding.cvCropImage.crop(builder.build())
    }

    override fun onCroppedRegionSaved(bitmapUri: Uri?) {
        dismiss()
        if (CropLibrary.isSaveToGalary) {
            val uri = BitmapUtil.saveBitmap2Galary(BitmapFactory.decodeFile(bitmapUri!!.path), this)
            CropLibrary.galleryAddPic(this@CropActivity, uri)
        }
        val intent = Intent()
        intent.putExtra(CropLibrary.EXTRA_CROP_ITEM_OUT_URI, bitmapUri)
        setResult(Activity.RESULT_OK, intent)
        finish()
    }

    override fun onError(var1: Throwable?) {
        dismiss()
    }

    private fun dismiss() {
        if (mProgressDialog != null && mProgressDialog!!.isShowing) {
            mProgressDialog!!.dismiss()
        }

    }

    override fun onDestroy() {
        super.onDestroy()
        dismiss()
        mProgressDialog = null
    }

    private fun releaseFolder() {
        FileUtils.releaseFolder(CropLibrary.CACHE_IMG_CROP!!)
    }

    companion object {
        const val REQUEST_CODE_ITEM_CROP = 3001
        const val EXTRA_URI = "CropImage"

        fun start(activity: AppCompatActivity, imageUri: Uri) {
            val intent = Intent(activity, CropActivity::class.java)
            intent.putExtra(EXTRA_URI, imageUri)
            activity.startActivityForResult(intent, REQUEST_CODE_ITEM_CROP)
        }
    }
}
